doc: add engineering overall file (#1524)
argos-ci/argos: docs/engineering.md

# Argos Engineering Documentation & Technical References

> This document outlines Argos’ architecture, explains how it’s hosted on Heroku using AWS S3 for image storage, and describes where to find further engineering references. The Argos CLI is maintained in a separate repository at [argos-ci/argos-javascript](https://github.com/argos-ci/argos-javascript).

Reviewer: ```suggestion
This document outlines Argos’ architecture, explains how it’s hosted on Heroku using AWS S3 for image storage, and describes where to find further engineering references. The Argos SDK are maintained in a separate repository at [argos-ci/argos-javascript](https://github.com/argos-ci/argos-javascript).
```

## Table of Contents

- [Overview](#overview)
- [Hosting & Storage](#hosting--storage)
- [Architecture Highlights](#architecture-highlights)
- [Development Workflow](#development-workflow)
- [Technical References](#technical-references)

---

## Overview

Argos is a visual testing platform that helps teams:

- Compare screenshots
- Detect UI regressions quickly
- Integrate seamlessly into various CI workflows

> It achieves this by hosting its core application on Heroku, uploading screenshots to AWS S3, and providing a companion CLI to facilitate automation from different CI services.

Reviewer: ```suggestion
It achieves this by hosting its core application on Heroku, uploading screenshots to AWS S3, and providing some SDKs to facilitate automation from different CI services and testing frameworks.
```

---

## Development Workflow

- **Install & Setup & Testing**
  Follow the instructions in the [CONTRIBUTING.md](../CONTRIBUTING.md) file to set up the development environment and run tests.

- **Deployment**
  - **Heroku**: Automatically triggers `pnpm run heroku-postbuild` (which in turn runs `build`) before deploying.
  - **Production Builds**: `pnpm run build` compiles and optimizes both backend and frontend assets.

---

## Technical References

Argos relies on a modern, TypeScript-heavy toolchain and documentation scattered across multiple files and folders:

1. **Core Configuration & Scripts**

   - **`package.json`**: Defines top-level scripts (e.g., `dev`, `build`, `test`), pinned Node version, and the main dev dependencies.
   - **`turbo.json`**: Coordinates concurrent build/test steps in a monorepo environment.
   - **`.env.example`**: Shows typical environment variables for local or Heroku setups.

2. **Backend Code**

   - Written in TypeScript.
   - Uses **knex** for database migrations and queries.
   - Interacts with AWS S3 for screenshot storage.

>    - GraphQL-related modules leverage **@graphql-codegen** for generating typed schema/resolvers.

Reviewer: ```suggestion
   - GraphQL-related modules leverage **graphql-codegen** for generating typed schema/resolvers.
```

3. **Frontend Code**

   - A React-based SPA that uses routes, components, and various hooks.
   - Configured to handle screenshot diff results and present build statuses.
   - Tailwind CSS for quick UI styling.

4. **Testing Infrastructure**

   - **Vitest** for unit and integration tests.
   - **Playwright** for end-to-end testing.

> 5. **CLI Tool**

Reviewer: ```suggestion
5. **SDKs**
```
